streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -817,14 +867,17 @@ private boolean handleRestoredTasksFromStateUpdater(final long now,
         for (final Task task : stateUpdater.drainRestoredActiveTasks(timeout)) {
             Set<TopicPartition> inputPartitions;
             if ((inputPartitions = tasks.removePendingTaskToRecycle(task.id())) != null) {
-                recycleTask(task, inputPartitions, tasksToCloseDirty, taskExceptions);
+                recycleTaskFromStateUpdater(task, inputPartitions, tasksToCloseDirty, taskExceptions);
             } else if (tasks.removePendingTaskToCloseClean(task.id())) {
                 closeTaskClean(task, tasksToCloseDirty, taskExceptions);
             } else if (tasks.removePendingTaskToCloseDirty(task.id())) {
                 tasksToCloseDirty.add(task);
             } else if ((inputPartitions = tasks.removePendingTaskToUpdateInputPartitions(task.id())) != null) {
                 task.updateInputPartitions(inputPartitions, topologyMetadata.nodeToSourceTopics(task.id()));
                 transitRestoredTaskToRunning(task, now, offsetResetter);
+            } else if (tasks.removePendingActiveTaskToSuspend(task.id())) {
+                task.suspend();
+                tasks.addTask(task);
```

Comment:
Here we handle revoked active tasks that have been already restored.

Reply:
We still need the set of pending active tasks to suspend because an iteration of the poll loop might happen between the calls to `handleRevocation()` and `handleAssignment()`. During that iteration a revoked active task might be removed from the state updater because it is restored and processed which would violate the assumption that revoked tasks are not processed (i.e., commit offsets) anymore.

Reply:
Ah right, that makes sense. We'd still need the pending suspend tasks then.